Use a hardcoded seed value for the bloom filter

ptschip opened the pull request:
When running thinblock_tests.cpp there is occassionally a failure
due to getting a false positive in the bloom filter. Using a hard
coded seed will alwasy give us the same result for the test and
prevent the failures.

gandrewstone commented on src/unlimited.cpp:
```diff
@@ -1382,7 +1382,11 @@ void BuildSeededBloomFilter(CBloomFilter& filterMemPool, std::vector<uint256>& v
+    if (Params().NetworkIDString() == "main")
+        filterMemPool = CBloomFilter(nElements, nFPRate, insecure_rand(), BLOOM_UPDATE_ALL);
+    else
+        filterMemPool = CBloomFilter(nElements, nFPRate, 5, BLOOM_UPDATE_ALL);
```
whenever we split what we test from what we run in the code, I get worried.  This seems to be very innocuous, but there are always possible issues like what if somebody makes a subsequent change to just one of these two lines by accident?

Is there any way we can push this into the test itself?  Maybe make it tolerant of false positives?

ptschip commented:
I agree, i don't really like doing that either...I wonder whether we
really need to do any random seeing anyway from a security standpoint,
but let me think about it for a while, maybe I can play with the filter
size or false positive rate.

28/12/2016 6:21 PM, gandrewstone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/unlimited.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/195#pullrequestreview-14626533>:
>
> > @@ -1382,7 +1382,11 @@ void BuildSeededBloomFilter(CBloomFilter& filterMemPool, std::vector<uint256>& v
>      if (nTimePassed > nHoursToGrow * 3600)
>          nFPRate = nMaxFalsePositive;
>  
> -    filterMemPool = CBloomFilter(nElements, nFPRate, insecure_rand(), BLOOM_UPDATE_ALL);
> +    // Use a hardcoded seed value when not on mainnet so that we don't get false positives when running thinblock_tests.cpp
> +    if (Params().NetworkIDString() == "main")
> +        filterMemPool = CBloomFilter(nElements, nFPRate, insecure_rand(), BLOOM_UPDATE_ALL);
> +    else
> +        filterMemPool = CBloomFilter(nElements, nFPRate, 5, BLOOM_UPDATE_ALL);
>
> whenever we split what we test from what we run in the code, I get
> worried. This seems to be very innocuous, but there are always
> possible issues like what if somebody makes a subsequent change to
> just one of these two lines by accident.
>
> Is there any way we can push this into the test itself? Maybe make it
> tolerant of false positives?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/195#pullrequestreview-14626533>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0NotLp3pwfq4mIW4BYBosgcx-EF4ks5rMxkugaJpZM4LW95e>.
>

ptschip committed:
Make it possible to create Deterministic bloom filters for thinblocks

This will prevent any false positives from happening when  unit
tests are run which then gives a failed test result when it
wasn't a true failure.

Add non-deterministic tests to thinblock_tests.cpp

ptschip commented:
@gandrewstone  OK, i took a different approach, i think this is better.
